### ISSUE TITLE
Metacontroller CRD's are in protected group

### ISSUE THREAD

**Issue by grzesuav** [metacontroller/metacontroller#54]
Thanks @fpetkovski for spotting,

metacontroller crd's are in group called `metacontroller.k8s.io`, which mean it is protected - https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190612-crd-group-protection.md#k8sio-group-protection

so we temporary need to add annotations marking it is unapproved - https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190612-crd-group-protection.md#what-to-do-if-you-accidentally-put-an-unapproved-api-in-a-protected-group - as in https://github.com/metacontroller/metacontroller/pull/53/

# Final solution

Either we need to ask for approval https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190612-crd-group-protection.md#what-to-do-if-you-accidentally-put-an-unapproved-api-in-a-protected-group or we should change the group name.

I would firstly check if our api is approvable,and what requirements we need to met.

**Comment by AmitKumarDas** [comment]
How about using `metacontroller.github.com` as the API group in the CRDs exposed in this project?
We can change the go mod to point this as well?
cc @grzesuav @fpetkovski

**Comment by grzesuav** [comment]
> How about using `metacontroller.github.io` as the API group in the CRDs exposed in this project?

To be honest it will cause not compatible change for API. If it will be necessary we must provide seamless migration path which will not cause breaks in operations.

> We can change the `go.mod` to point this as well?

Which part of `go.mod` you have in mind ?

**Comment by fpetkovski** [comment]
I also have concerns that changing the API group will cause a backwards incompatible change and break existing users if not done carefully. The `group` field in the CRD spec is immutable, so a migration from existing users would need to include:
- deploying a new version of metacontroller that watches `*.metacontroller.github.io`
- applying CRDs with a new name in the `metacontroller.github.io` group
- applying `CompositeController`s and `DecoratorController`s from the apiGroup `metacontroller.github.io`
- removing CRDs from the old group.

Based on a basic investigation I think that it could be done without downtime, but it also involves taking some risks.

**Comment by grzesuav** [comment]
> I also have concerns that changing the API group will cause a backwards incompatible change and break existing users if not done carefully. The `group` field in the CRD spec is immutable, so a migration from existing users would need to include:
> 
> * deploying a new version of metacontroller that watches `*.metacontroller.github.io`
> * applying CRDs with a new name in the `metacontroller.github.io` group
> * applying `CompositeController`s and `DecoratorController`s from the apiGroup `metacontroller.github.io`
> * removing CRDs from the old group.
> 
> Based on a basic investigation I think that it could be done without downtime, but it also involves taking some risks.

agree, I guess in 1.16 there were introduced some webhook conversion API, but anyway, if it is possible, it would be good to not doing that at all.

**Comment by grzesuav** [comment]
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion

**Comment by AmitKumarDas** [comment]
Alright. Migration will be painful for sure. Let's ask for approval then. @grzesuav or @fpetkovski do you mind taking this up.

Eventually I am hoping we can make this CRD stuff an internal affair similar to a configuration file.  I have had good success in metac in this approach. However getting an approval should be the right thing for existing CRDs.

**Comment by grzesuav** [comment]
Sure, I wanted to take it after merging some improvements to releases, testing @fpetkovski changes (CRD change). Will try to do some discovery next week

**Comment by fpetkovski** [comment]
I also did some preliminary research and managed to find the process required for getting an API approved: https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md#mechanics

**Comment by grzesuav** [comment]
thanks @fpetkovski !

In first place we should look at :
* https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md
* https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md

to check if we comply
